daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Anf.scala

```diff
@@ -0,0 +1,339 @@
+    val depth = DepthA(0)
+    val env = initEnv
+    flattenExp(depth, env, exp, flattenedExpression => Land(flattenedExpression)).bounce
+  }
+
+  /**
+    The transformation code is implemented using a continuation-passing style of
+    translation (which is quite common for translation to ANF). In general, naming nested
+    compound expressions requires turning an expression kind of inside out, lifting the
+    introduced let-expression up to the the nearest enclosing abstraction or case-branch.
+
+    For speedy, the ANF pass occurs after translation to De Brujin and closure conversions,
+    which adds the additional complication of re-indexing the variable indexes. This is
+    achieved by tracking the old and new depth & the mapping between them. See the types:
+    DepthE, DepthA and Env.
+
+    There is also the issue of avoiding stack-overflow during compilation, which is
+    managed by the using of a Trampoline[T] type.
+    */
+  case class CompilationError(error: String) extends RuntimeException(error)
+
+  /** `DepthE` tracks the stack-depth of the original expression being traversed */
+  case class DepthE(n: Int)
+
+  /** `DepthA` tracks the stack-depth of the ANF expression being constructed */
+  case class DepthA(n: Int)
+
+  /** `Env` contains the mapping from old to new depth, as well as the old-depth as these
+    * components always travel together */
+  case class Env(absMap: Map[DepthE, DepthA], oldDepth: DepthE)
+
+  val initEnv = Env(absMap = Map.empty, oldDepth = DepthE(0))
+
+  def trackBindings(depth: DepthA, env: Env, n: Int): Env = {
+    if (n == 0) {
+      env
+    } else {
+      val extra = (0 to n - 1).map(i => (DepthE(env.oldDepth.n + i), DepthA(depth.n + i)))
+      Env(absMap = env.absMap ++ extra, oldDepth = DepthE(env.oldDepth.n + n))
+    }
+  }
+
+  // TODO: reference something here about trampolines
+  sealed abstract class Trampoline[T] {
+    @tailrec
+    final def bounce: T = this match {
+      case Land(x) => x
+      case Bounce(continue) => continue().bounce
+    }
+  }
+
+  final case class Land[T](x: T) extends Trampoline[T]
+  final case class Bounce[T](continue: () => Trampoline[T]) extends Trampoline[T]
+
+  /** `Res` is the final, fully transformed ANF expression, returned by the continuations. */
+  type Res = Trampoline[AExpr]
+
+  /** `K[T]` is the continuation type which must be passed to the core transformation
+    functions, i,e, `transformExp`.
+
+    Notice how the DepthA is threaded through the continuation.
+    */
+  type K[T] = ((DepthA, T) => Res)
+
+  /** During conversion we need to deal with bindings which are made/found at a given
+    absolute stack depth. These are represented using `AbsBinding`.
+
+    Note the contrast with the expression form `ELocS` which indicates a relative offset
+    from the top of the stack. This relative-position is used in both the original
+    expression which we traverse AND the new ANF expression we are constructing.
+    */
```

Comment:
I've improved the explanation in the comment,

Reply:
> An absolute stack depth is the offset from the bottom of the stack, which is empty when a function is entered.

That is not the case because we use the same stack throughout the whole interpretation. I would prefer to clarify this even further.

Reply:
clarified further